src/dtree.ml

```diff
@@ -0,0 +1,648 @@
+(** This module provides functions to compile rewrite rules to decision trees
+    in order to pattern match the rules efficiently.  The method is based on
+    Luc Maranget's {i Compiling Pattern Matching to Good Decision Trees},
+    {{:10.1145/1411304.1411311}DOI}. *)
+open Terms
+open Extra
+module Sub = Basics.Subterm
+module SubMap = Basics.SubtMap
+
+(** See {!type:tree} in {!module:Terms}. *)
+type t = tree
+
+(** Type of the leaves of the tree.  See {!module:Terms}, {!field:rhs}. *)
+type action = (term_env, term) Bindlib.mbinder
+
+(** {b Example} Given a rewrite system for a symbol [f] given as
+    - [f Z (S m)     → S m]
+    - [f n Z         → n]
+    - [f (S n) (S m) → S (S m)]
+
+    A possible tree might be
+    {v
++–?–∘–Z–∘     → n
+├–Z–∘–Z–∘     → n
+|   └–S–∘–?–∘ → S m
+└–S–∘–Z–∘     → n
+    └–S–∘–?–∘ → S (S m)
+    v}
+    with [∘] being a node (with a label not shown here) and [–u–]
+    being an edge with a matching on symbol [u] or a variable or wildcard when
+    [?].  Typically, the portion [S–∘–Z] is made possible by a swap. *)
+
+(** {3 Reduction substrate} *)
+
+(** Element consumed when reducing terms.  When reducing, we must have
+    - fast access to any element in the substrate (for swaps)
+    - fast replacement of an element and extension to replace an element of
+      the substrate by its reduced form, or an unfolding of
+      {!constructor:Appl} nodes.
+
+    Arrays have fast access but can't be extended efficiently.  Lists can be
+    extended easily but access is slow.  Tree structure have moderately fast
+    access and extension, and are therefore appropriate.
+
+    Two implementations have been tried with
+    - {!module:BatVect},
+    - {!module:BatFingerTree}
+    The performances are similar. *)
+module type reduction_substrate =
+sig
+  (** Type of a substrate of ['a]. *)
+  type 'a t
+
+  (** The empty substrate. *)
+  val empty : 'a t
+
+  (** [is_empty v] returns whether a stack is empty. *)
+  val is_empty : 'a t -> bool
+
+  (** [of_list l] returns a stack containing the elements of [l]. *)
+  val of_list : 'a list -> 'a t
+
+  (** [length v] is the number of elements in [v]. *)
+  val length : 'a t -> int
+
+  (** Prefix and suffix of the substrate. *)
+  type 'a prefix
+  type 'a suffix
+
+  (** [destruct v i] returns a triplet [(r, e, o)] with [r] being the
+      elements from 0 to [i - 1], [e] the [i]th element and [o] the elements
+      from [i + 1] to the end of [v]. *)
+  val destruct : 'a t -> int -> 'a prefix * 'a * 'a suffix
+
+  (** [restruct r n o] is the concatenation of three stacks [r] [n] and
+      [o]. *)
+  val restruct : 'a prefix -> 'a list -> 'a suffix -> 'a t
+end
+
+module ReductionStack : reduction_substrate =
```

Comment:
Please add a comment saying that it is an implementation based on lists.
So, why not calling it RedStackList for instance, and later define RedStack as RedStackList?

Reply:
You could also add a comment on the complexity of each function. Note that length is linear. Is length called often? This may be improved by storing the length with the list.

Reply:
Length is not the costlier function when profiling, but I might try (@rlepigre already tried, and it ended not that well)

Reply:
I tried again, it is slower as `destruct` and `restruct` are the slower functions; and both are heavier when using constant time length structures